### ISSUE TITLE
TensorForest: the base random seed takes no effects

### ISSUE THREAD

**Issue by whatbeg** [tensorflow/tensorflow#24040]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary, pip install tensorflow==1.12.0
- TensorFlow version (use command below): 1.12.0
- Python version: 3.6.4
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: No usage
- GPU model and memory: No usage

**Describe the current behavior**

I found that setting the same base_random_seed for TensorForest generate different classification performance.

**Describe the expected behavior**

Setting the **same** base_random_seed for TensorForest should generate the **same** classification performance so that we can reproduce every experiment.

**Code to reproduce the issue**

```
import numpy as np
import tensorflow as tf
import sklearn.datasets
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

X, y = sklearn.datasets.load_digits(return_X_y=True)
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
x_train = x_train.astype(np.float32)
x_test = x_test.astype(np.float32)

params = tf.contrib.tensor_forest.python.tensor_forest.ForestHParams(
num_classes=10, num_features=64, regression=False,
num_trees=50, max_nodes=1000, base_random_seed=2)

classifier = tf.contrib.tensor_forest.client.random_forest.TensorForestEstimator(params)
classifier.fit(x=x_train, y=y_train)
y_out = classifier.predict(x=x_test)

yyy = [x for x in y_out]

y_hat = np.array([x['classes'] for x in yyy])
print(accuracy_score(y_test, y_pred=y_hat))
# first run:  0.9629629629629629
# second run: 0.968013468013468
# third run:  0.9646464646464646
```

The base random seed is non-zero, it's supposed to produce the same result.
Thus, I wonder where else does the code have randomness?
@yupbank

**Comment by ymodak** [comment]
The code snippet you provided looks incomplete. Can you please provide a complete repro code? Thanks!

**Comment by whatbeg** [comment]
@ymodak more complete : )

**Comment by whatbeg** [comment]
@yupbank hi, yupbank, can i ask you a question, is the TensorForest slow on small datasets? I trained on adult dataset (32561 samples, 113 features) and set 250 trees, max_nodes=10000, the trainning will cost 1200+ seconds. For this small dataset, I think this time cost is so large... especially compared to scikit-learn implementation. Was it my mis-use? 

```
x_train, x_test, y_train, y_test = load_adult()

start_time = time.time()
est_args = {'num_classes': 2, 'num_features': 113, 'regression': False,
            'num_trees': 250, 'max_leaf_nodes': 10000,
            'base_random_seed': 0}

params = tf.contrib.tensor_forest.python.tensor_forest.ForestHParams(**est_args)
estimator = tf.contrib.tensor_forest.client.random_forest.TensorForestEstimator(params)
estimator.fit(x_train, y_train)
print("Time Cost: {}".format(time.time() - start_time))

# Time Cost: 1282s
```

**Comment by yupbank** [comment]
hey... so there is already a hyperparam control the deterministic, try turn on `collate_examples=True`  in `ForestHParams` and the result would be reproducible

**Comment by whatbeg** [comment]
@yupbank oh, thank you very much. But i have another question, may i get your help ?
https://github.com/tensorflow/tensorflow/issues/24040#issuecomment-446980933, besides this time performance issue, I encountered another situation.
When I trained a TensorForestEstimator with 250 trees and max_nodes 10000, on a classification problem with num_classes=10, num_examples=4032930. I encounted a **ResourceExhaustedError**. as follows
```
ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[4032930,250,10] and type float on /job:localhost/replica:0/task:0/device:CPU:0 by allocator cpu
```
my machine: 64GB RAM and 12 cores.

Thus, I wonder, isn't it an online algorithm ? it seems that the memory consumption of the TensorForest is relatively large? How can I improve it ? 
Thanks : )

**Comment by yupbank** [comment]
thanks for the feedback, do you have a example code for me to reproduce the error? and you are right, it's a streaming algorithm, so there shouldn't be any memory problem.

and yes, i can take a look at the problem, but do you mind to create separate issues about the problems? so it's easier for people to share different piece of your valuable problems

and i think we can close this issue? feel free to reopen it..

**Comment by whatbeg** [comment]
> hey... so there is already a hyperparam control the deterministic, try turn on `collate_examples=True` in `ForestHParams` and the result would be reproducible

oh, what is the meaning of the hyperparam `collate_examples` ?
I set it to `True`, but it does not work ... 5555

My arguments are:
```
tf.set_random_seed(0)
#tf.logging.set_verbosity(tf.logging.INFO)
x_train, x_test, y_train, y_test = load_adult()
#x_train = x_train.reshape((60000, -1))

start_time = time.time()
est_args = {'num_classes': 2, 'num_features': 113, 'regression': False,
            'num_trees': 2, 'max_nodes': 1000, 'collate_examples': True,
            'base_random_seed': 0}
hparams = ForestHParams(**est_args)
estimator = ...
estimator.fit(x_train, y_trian)
y_out = estimator.predict(x_test)
```
The `y_out` differs every runs.
@yupbank

**Comment by yupbank** [comment]
Weird... are your using tf1.12 too? I can’t reproduce it

**Comment by whatbeg** [comment]
yes，1.12








----------------------------------------------------------
Best Wishes !
胡求（Qiu Hu）


2018-12-16

On 12/16/2018 16:45, Peng Yu wrote:
Weird... are your using tf1.12 too? I can’t reproduce it

**Comment by yupbank** [comment]
ah, you are right, i can only reproduce the results because my set the `max_nodes=10`

**Comment by yupbank** [comment]
i think the pr i linked would fix the problem, thanks for your feedback !

**Comment by mohantym** [comment]
Hi @whatbeg!, 
You seem to be using older versions of Tensorflow . We recommend that you upgrade your code base version to 2.x  as contrib is removed from 2.x versions and let us know if the issue still persists in newer versions. Thanks!
